server/store/middleware/infohash/blacklist.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2016 The Chihaya Authors. All rights reserved.
+// Use of this source code is governed by the BSD 2-Clause license,
+// which can be found in the LICENSE file.
+
+package infohash
+
+import (
+	"github.com/chihaya/chihaya"
+	"github.com/chihaya/chihaya/server/store"
+	"github.com/chihaya/chihaya/tracker"
+)
+
+func init() {
+	tracker.RegisterAnnounceMiddleware("infohash_blacklist", blacklistAnnounceInfohash)
+	tracker.RegisterScrapeMiddlewareConstructor("infohash_blacklist", blacklistScrapeInfohash)
+}
+
+// ErrBlockedInfohash is returned by a middleware if any of the infohashes
+// contained in an announce or scrape are disallowed.
+var ErrBlockedInfohash = tracker.ClientError("disallowed infohash")
+
+// blacklistAnnounceInfohash provides a middleware that only allows announces
+// for infohashes that are not stored in a StringStore
+func blacklistAnnounceInfohash(next tracker.AnnounceHandler) tracker.AnnounceHandler {
+	return func(cfg *chihaya.TrackerConfig, req *chihaya.AnnounceRequest, resp *chihaya.AnnounceResponse) (err error) {
+		blacklisted, err := store.MustGetStore().HasString(store.PrefixInfohash + string(req.InfoHash))
+
+		if err != nil {
+			return err
+		} else if blacklisted {
+			return ErrBlockedInfohash
+		}
+		return next(cfg, req, resp)
+	}
+}
+
+// blacklistScrapeInfohash provides a middleware constructor that returns a
+// a middleware that modifies scrape requests based on the infohashes scraped.
+//
+// The middleware works in two modes: block and filter.
+// The block mode blocks a scrape completely if any of the infohashes is
+// disallowed.
+// The filter mode filters any disallowed infohashes from the scrape,
+// potentially leaving an empty scrape.
+func blacklistScrapeInfohash(c chihaya.MiddlewareConfig) (tracker.ScrapeMiddleware, error) {
+	cfg, err := newConfig(c)
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.Mode == ModeBlock {
+		// block scrape if it contains a blacklisted infohash
+		return func(next tracker.ScrapeHandler) tracker.ScrapeHandler {
```

Comment:
pull these functions out and make this all inside a switch statement.

```go
switch cfg.Mode {
case ModeBlock:
  return blockMiddleware
case ModeFilter:
  return filterMiddleware
default:
  return filterMiddleware
}
```

It's much more understandable afterwards.